swupd/cmd/create-pack/main.go

```diff
@@ -0,0 +1,237 @@
+	// different directories, based on the version that each bundle is in the Manifest.MoM for
+	// the specified toVersion.
+	//
+	toDir := filepath.Join(stateDir, "www", toVersion)
+	toMoM, err := readManifestFile(filepath.Join(toDir, "Manifest.MoM"))
+	if err != nil {
+		log.Fatalf("couldn't read MoM of TO_VERSION (%s): %s", toVersion, err)
+	}
+
+	bundles := make(map[string]*bundleToPack, len(toMoM.Files))
+	for _, f := range toMoM.Files {
+		bundles[f.Name] = &bundleToPack{f.Name, 0, f.Version}
+	}
+
+	// If this is not a zero pack, we might be able to skip some bundles.
+	if fromVersionUint > 0 {
+		fromDir := filepath.Join(stateDir, "www", fromVersion)
+		fromMoM, err := readManifestFile(filepath.Join(fromDir, "Manifest.MoM"))
+		if err != nil {
+			log.Fatalf("couldn't read MoM of FROM_VERSION (%s): %s", fromVersion, err)
+		}
+
+		for _, f := range fromMoM.Files {
+			e, ok := bundles[f.Name]
+			if !ok {
+				// Bundle doesn't exist in new version, no pack needed.
+				continue
+			}
+			if e.to == f.Version {
+				// Versions match, so no pack required.
+				delete(bundles, f.Name)
+				continue
+			}
+			if e.to < f.Version {
+				log.Fatalf("invalid bundle versions for bundle %s, check the MoMs", f.Name)
+			}
+			e.from = f.Version
+		}
+	}
+
+	// TODO: Use goroutines.
+	for _, b := range bundles {
+		pack(stateDir, chrootDir, *outputDir, b, *force)
+	}
+}
+
+type bundleToPack struct {
+	name string
+	from uint32
+	to   uint32
+}
+
+func pack(stateDir, chrootDir string, outputDir string, bundle *bundleToPack, force bool) {
+	toDir := filepath.Join(outputDir, "www", fmt.Sprint(bundle.to))
+	err := os.MkdirAll(toDir, 0755)
+	if err != nil {
+		log.Fatal(err)
+	}
+	outputFilename := filepath.Join(toDir, fmt.Sprintf("pack-%s-from-%d.tar", bundle, bundle.from))
+
+	// If we are not forcing, skip the packs already on disk.
+	if !force {
+		_, err := os.Stat(outputFilename)
+		if err == nil {
+			fmt.Printf("Pack already exists for %s from %d to %d\n", bundle.name, bundle.from, bundle.to)
+			return
+		}
+		if !os.IsNotExist(err) {
+			log.Fatal(err)
+		}
+	}
+
+	fmt.Printf("Packing %s from %d to %d...\n", bundle.name, bundle.from, bundle.to)
+
+	m, err := readManifestFile(filepath.Join(stateDir, "www", fmt.Sprint(bundle.to), "Manifest."+bundle.name))
+	if err != nil {
+		log.Fatal(err)
+	}
+	m.Name = bundle.name
+
+	//
+	// Create the file and write the pack to it.
+	//
+	output, err := os.Create(outputFilename)
+	if err != nil {
+		log.Fatal(err)
+	}
+	info, err := swupd.WritePack(output, m, bundle.from, filepath.Join(stateDir, "www"), chrootDir)
+	if err != nil {
+		os.RemoveAll(outputFilename)
+		log.Fatal(err)
+	}
+	err = output.Close()
+	if err != nil {
+		os.RemoveAll(outputFilename)
+		log.Fatal(err)
+	}
+
+	//
+	// Output information about the pack produced.
+	//
+	for i, w := range info.Warnings {
+		if i == 0 {
+			fmt.Println("Warnings during pack:")
+		}
+		fmt.Printf("  %s\n", w)
+		if i == len(info.Warnings)-1 {
```

Comment:
Instead of checking here, why not just do the `fmt.Println()` after the loop exits?

Reply:
I just want the extra newline if there were warnings.

Reply:
Ended up adding a if block around the loop and the prints.